src/infrastructure/storage.py

```
@@ -0,0 +1,179 @@
+        except Exception as e:
+            raise CheckpointException(f"Failed to save checkpoint: {e}")
+
+    async def close(self) -> None:
+        await self.flush()
+        if self._file_handle:
+            await self._file_handle.close()
+            self._file_handle = None
+
+    async def _count_existing_rows(self) -> int:
+        count = 0
+        async with aiofiles.open(self.output_file, 'r') as f:
```

Ruff (0.12.2)

Unnecessary mode argument

Remove mode argument

(UP015)